### PR TITLE
Proper Subnormal Rounding When Interpreting

### PR DESCRIPTION
**Pull request by Geotale**
During 25-bit rounding, subnormals are "normalized", which would normally mean that the exponent needs to be able to be lower than -1023 (which is what PPC does internally), but in actuality it's possible to simply modify at what bit rounding occurs and get the same results!
This likely hasn't been an issue anywhere due to the odd requirements for it to be relevant -- You need to run some single-precision multiplication instruction, have an extremely large frA, and have a double precision subnormal in frC, which shouldn't ever get generated by any compilers (due to it throwing double precision values into a single precision instruction), but does improve the overall accuracy of the interpreter!!
This is also why it might be best not to add it into the JITs, adding a performance penalty in order to accommodate for a case which should and likely never does happen ^^;

Tested on some ghosts in Mario Kart Wii to verify sync (despite their unlikelihood), along with handcrafted tests of a few cases, including one prevalent in a PPC test program which multiplies the bits `0x7fe0000000000000` with the certain subnormal powers of 2 - 1, also verifying the rounding behavior works properly!
Expected results verified on a Wii U executing in Wii mode

### CODE REVIEW

**Comment by Pokechu22** [comment]
Thanks for your contribution! I don't have time to review it right now, but I wanted to point out that we have some automated tests of floating-point stuff at https://github.com/dolphin-emu/hwtests/tree/master/cputest. If you have some test code for this already written it'd be nice to include this there.

**Comment by Geotale** [comment]
Can do! :O

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_FPUtils.h [hunk 1]
```diff
@@ -92,7 +92,33 @@ inline double Force25Bit(double d)
...
+    // at a different bit
+
+    s64 keep_mask = 0xFFFFFFFFF8000000LL;
+    u64 round = 0x8000000;
+
+    // Shift the mask and rounding bit to the right until
+    // the fraction is "normal"
+    // That is to say shifting it until the MSB of the fraction
+    // would escape into the exponent
+    u32 shift = std::countl_zero(fraction) + Common::DOUBLE_FRAC_WIDTH - 63;
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/Interpreter/Interpreter_FPUtils.h:111]: Just a nitpick, but personally I find this easier to reason about if written like:

```suggestion
    u32 shift = std::countl_zero(fraction) - (63 - Common::DOUBLE_FRAC_WIDTH);
```

**Review by JosJuice — COMMENTED** [review]
Thanks for finding this and writing a fix!

I was curious if this might fix the desync between console and Dolphin in 1080° Snowboarding, so I tested it (using the JIT, with all float and paired instructions falling back to the interpreter). Unfortunately it didn't help.

> This is also why it might be best not to add it into the JITs, adding a performance penalty in order to accommodate for a case which should and likely never does happen ^^;

Agreed. There's actually another reason why it shouldn't happen: Games usually set the non-IEEE bit, which means floating-point operations never generate subnormals.

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_FPUtils.h (hunk 1 above)
Comment (Geotale) [reply Source/Core/Core/PowerPC/Interpreter/Interpreter_FPUtils.h:111]: That's understandable :o Both make perfect sense I think ^^

**Comment by Geotale** [comment]
There are plenty more floating point things to fix in Dolphin, as weird as it seems ^^ I'd love to help in any way I can to improve the accuracy of my other things, and still have quite a few still up my sleeve :D (this CPU test program covers a *lot* ^^; Dolphin's interpreter still fails 325 tests or so iirc...)

**Comment by JosJuice** [comment]
LGTM if you squash the commits into one.

**Comment by Geotale** [comment]
I haven't gotten hwtests set up still, but I did squash the commits :>

**Comment by JosJuice** [comment]
The commits don't look squashed to me. Did you make sure to push?

**Comment by JMC47** [comment]
Ping about squashing the commits into one so this can get merged.

**Comment by JosJuice** [comment]
Please change the commit title so it matches the PR title. "Correct Removal of Loop [squash]" sounds like the name of one of the squashed commits, not of the larger change.

**Comment by JosJuice** [comment]
That's still not right. "Correct Removal of Loop" doesn't match what the PR is called or what the change does.

**Comment by Geotale** [comment]
You are *right* I am so sorry;;

**Pull request merged**